fpdf/fonts.py

```diff
@@ -131,20 +156,81 @@ def __init__(self, fpdf, font_file_path, fontkey, style):
             if w == 65535:
                 w = 0
 
-            self.cw[char] = round(scale * w + 0.001)  # ROUND_HALF_UP
+            self.cw[char] = round(self.scale * w + 0.001)  # ROUND_HALF_UP
+
+            self.glyph_ids[char] = font.getGlyphID(glyph)
 
         # include numbers in the subset! (if alias present)
         # ensure that alias is mapped 1-by-1 additionally (must be replaceable)
-        sbarr = "\x00 "
+        sbarr = "\x00 \n\r"
         if fpdf.str_alias_nb_pages:
             sbarr += "0123456789"
             sbarr += fpdf.str_alias_nb_pages
 
         self.name = re.sub("[ ()]", "", font["name"].getBestFullName())
-        self.up = round(font["post"].underlinePosition * scale)
-        self.ut = round(font["post"].underlineThickness * scale)
+        self.up = round(font["post"].underlinePosition * self.scale)
+        self.ut = round(font["post"].underlineThickness * self.scale)
         self.emphasis = TextEmphasis.coerce(style)
-        self.subset = SubsetMap([ord(char) for char in sbarr])
+        self.subset = SubsetMap(self, [ord(char) for char in sbarr])
+
+    def get_text_width(self, text, font_size_pt, text_shaping):
+        if text_shaping:
+            return self.shaped_text_width(text, font_size_pt)
+        return (len(text), sum(self.cw[ord(c)] for c in text) * font_size_pt * 0.001)
+
+    def shaped_text_width(self, text, font_size_pt):
+        if hb == None:
+            raise FPDFException(
+                'uharfbuzz wasn\'t imported. Use "pip install uharfbuzz"'
+            )
+        if not hasattr(self, "hbfont"):
+            self.hbfont = hb.Font(hb.Face(hb.Blob.from_file_path(self.ttffile)))
+        self.hbfont.ptem = font_size_pt
+        buf = hb.Buffer()
+        buf.add_str("".join(text))
+        buf.guess_segment_properties()
+        features = {"kern": True, "liga": True}
+        hb.shape(self.hbfont, buf, features)
+        text_width = 0
+        for pos in buf.glyph_positions:
+            text_width += pos.x_advance
+            text_width += pos.x_offset
+        return (len(buf.glyph_positions), text_width)
+
+    def convert_pdf_text(self, text):
```

Comment:
```suggestion
    def encode_text(self, text):
```
It's great that you refactored this.
I'm looking for a more explicit method name here.
What do you think of this one? Do you have other suggestions.